### PR TITLE
Allow repaint to occur before removing ripple visibility

### PR DESCRIPTION
**Pull request by cainmi**
Resolves google/material-design-lite#376

### CODE REVIEW

**Comment by cainmi** [comment]
I signed it!

File: src/ripple/ripple.js [hunk 1]
```diff
@@ -118,7 +119,12 @@ MaterialRipple.prototype.upHandler_ = function(event) {
 
   // Don't fire for the artificial "mouseup" generated by a double-click.
   if (event && event.detail !== 2) {
-    this.rippleElement_.classList.remove(this.CssClasses_.IS_VISIBLE);
+    // Allow a repaint to occur before removing this class, so the animation
+    // shows for tap events, which seem to trigger a mouseup too soon after
+    // mousedown.
+    window.setTimeout(function() {
```
Comment (samccone) [inline src/ripple/ripple.js]: I think we will want to use the "next-tick" approach here much like react-animation does..
I am pretty sure we can just use setTimeout() 0 to defer this class toggle on the stack. Perhaps there is a more elegant solution here.

Thanks @cainmi :+1:

Comment (cainmi) [inline src/ripple/ripple.js]: I originally used `setTimeout(fn, 0)`, but I noticed `TINY_TIMEOUT` used a fair bit in other components (possibly for similar issues), so wanted to keep things consistent.

Comment (addyosmani) [inline src/ripple/ripple.js]: Thanks for keeping things consistent, @cainmi. LGTM. @samccone do you have a link to a react-animation line that's approaching this using next-tick?

Comment (surma) [inline src/ripple/ripple.js]: Is “next-tick” the same as `requestAnimationFrame`? As we are polyfilling rAF anyways, we might as well use it.

Comment (samccone) [inline src/ripple/ripple.js]: Ok so in the react-animation docs they mention a "next-tick" approach..
https://github.com/facebook/react/blame/master/docs/docs/10.1-animation.md#L59

After a bit of digging into the implementation details here is the line in question

https://github.com/facebook/react/blob/34c8b2366ca4137b0a59f679fa12ae676da76119/src/addons/transitions/ReactCSSTransitionGroupChild.js#L95-L101

So it looks like `TICK` is a magical 17.... :rabbit2: for their setTimeout

So to be clear, the current approach is inline with the next-tick approach :gem: , but there is still a question of the correct timeout and being consistent internally. It has been my understanding that a timeout of 0 will accomplish the goal here of dequeuing and enqueuing the task to the bottom of the stack... but perhaps there is something here..... :cactus:

Comment (cainmi) [inline src/ripple/ripple.js]: I was under the impression that a timeout of 0 was sufficient as well, but for the sake of consistency maybe just merge this PR, and then drop all instances of `TINY_TIMEOUT` in another commit, and abstract out all the individual `setTimeout` calls to a shared helper method.

I think you can actually do `setTimeout(fn)`, i.e. ommit the second argument, and it defaults to 0.

Comment (cainmi) [inline src/ripple/ripple.js]: Given that the second (delay) argument in `setTimout` is in milliseconds, I doubt 0.0001 is much different to 0 as far as browsers are concerned.

Comment (samccone) [inline src/ripple/ripple.js]: As I was walking around this evening I had somewhat of a revelation.

60FPS === a tick every 16.666.... MS ... so 17 becomes way less magical if react was optimizing here for a 60FPS system (where at most we could paint 60 frames a secong).

Given that knowledge, I would argue that TINY_TIMEOUT should actually be `17` 

:cool:

Comment (cainmi) [inline src/ripple/ripple.js]: I think we might be merging two issues here...
In this particular case we aren't looking at frames, we just want to push the function to the end of the execution queue.

**Comment by surma** [comment]
> In this particular case we aren't looking at frames
Ah, from the PR title I assumed we were actually looking for a proper repaint, which is why I thought `requestAnimationFrame` would be more appropriate. If we are not looking for a repaint, `setTimeout(...,0)` should do fine (although I still like the semantic value of rAF.

@cainmi Could you test which of these two approaches actually work?

File: src/ripple/ripple.js (hunk 1 above)
Comment (surma) [inline src/ripple/ripple.js]: > In this particular case we aren't looking at frames

Ah, from the PR title I assumed we were actually looking for a proper repaint, which is why I thought `requestAnimationFrame` would be more appropriate. If we are not looking for a repaint, `setTimeout(...,0)` should do fine (although I still like the semantic value of rAF.

@cainmi Could you test which of these two approaches actually work?

Comment (cainmi) [inline src/ripple/ripple.js]: Looks like both work.

Comment (surma) [inline src/ripple/ripple.js]: @addyosmani @samccone WDYT? I’d actually vote for `setTime(...,0)` for clarity’s sake.

Comment (addyosmani) [inline src/ripple/ripple.js]: setTimeout imo, but I'm fine with either.

Comment (samccone) [inline src/ripple/ripple.js]: I would perhaps in a pedantic world :fire: :moon:  argue for RAF or a setTimeout of 17 ... but this is not a big deal at all :) It was a fun learning experience :books:

Comment (surma) [inline src/ripple/ripple.js]: @samccone I’d agree, but we are not actually waiting for the next frame, we just want to push some code back on the queue so we can yield to the browser.

@cainmi Would you mind changing to `setTimeout(...,0)`, than this is good to go :) Thanks for your patience, this blew up a little :boom:

**Comment by addyosmani** [comment]
@surma feel free to merge once that one last change is made :) @cainmi thank you so much for all your help here :star2: :dizzy:

**Comment by samccone** [comment]
:clap: thanks again @cainmi :a: :heavy_plus_sign:

**Comment by cainmi** [comment]
No worries, ready to merge.

**Comment by surma** [comment]
Let’s do this! :shipit:

**Pull request merged**